modules/internal/Atomics.chpl

```diff
@@ -278,148 +101,205 @@ module Atomics {
+                else return "atomic_" + s + "_"          + externTString(T);
+  }
+
   pragma "no doc"
-  proc chpl__processorAtomicType(type base_type) type {
-    if base_type==bool then return atomicbool;
-    else if base_type==uint(8) then return atomic_uint8;
-    else if base_type==uint(16) then return atomic_uint16;
-    else if base_type==uint(32) then return atomic_uint32;
-    else if base_type==uint(64) then return atomic_uint64;
-    else if base_type==int(8) then return atomic_int8;
-    else if base_type==int(16) then return atomic_int16;
-    else if base_type==int(32) then return atomic_int32;
-    else if base_type==int(64) then return atomic_int64;
-    else if base_type==real(64) then return atomic_real64;
-    else if base_type==real(32) then return atomic_real32;
-    else compilerError("Unsupported atomic type");
+  proc chpl__processorAtomicType(type T) type {
+    if T == bool           then return AtomicBool;
+    else if isSupported(T) then return AtomicT(T);
+    else compilerError("Unsupported atomic type: " + T:string);
   }
 
+  // Parser hook
   pragma "no doc"
-  proc chpl__atomicType(type base_type) type {
+  proc chpl__atomicType(type T) type {
     if CHPL_NETWORK_ATOMICS == "none" {
-      return chpl__processorAtomicType(base_type);
+      return chpl__processorAtomicType(T);
     } else {
-      return chpl__networkAtomicType(base_type);
+      return chpl__networkAtomicType(T);
     }
-  };
-
+  }
 
   pragma "atomic type"
   pragma "ignore noinit"
-  /*
-     The boolean atomic type.
-  */
-  record atomicbool {
+  record AtomicBool {
     pragma "no doc"
-    var _v:atomic_bool;
+    var _v:externT(bool);
 
     pragma "no doc"
     proc init() {
+      extern externFunc("init", bool, explicit=false)
+        proc atomic_init(ref obj:externT(bool), value:bool): void;
+
       this.complete();
-      atomic_init_bool(_v, _defaultOf(bool));
+      atomic_init(_v, _defaultOf(bool));
     }
 
     pragma "no doc"
     proc deinit() {
-      atomic_destroy_bool(_v);
+      extern externFunc("destroy", bool, explicit=false)
+        proc atomic_destroy(ref obj:externT(bool)): void;
+
+      atomic_destroy(_v);
     }
 
     /*
        :returns: The stored value.
     */
-    inline proc const read(order:memory_order = memory_order_seq_cst):bool {
+    inline proc const read(order:memory_order = memory_order_seq_cst): bool {
+      extern externFunc("load", bool)
+        proc atomic_load(const ref obj:externT(bool), order:memory_order): bool;
+
       var ret:bool;
-      on this do ret = atomic_load_explicit_bool(_v, order);
+      on this do ret = atomic_load(_v, order);
       return ret;
     }
 
     /*
        Stores `value` as the new value.
     */
-    inline proc write(value:bool, order:memory_order = memory_order_seq_cst) {
-      on this do atomic_store_explicit_bool(_v, value, order);
+    inline proc write(value:bool, order:memory_order = memory_order_seq_cst): void {
```

Comment:
We've talked about making these memory_order arguments param. Do you think that's a good idea? Do you know if we have an issue about it on the Backlog somewhere?

Reply:
Yeah, part of https://github.com/chapel-lang/chapel/issues/10001